o1vm/src/pickles/proof.rs

```diff
@@ -1,39 +1,60 @@
+use std::collections::BTreeMap;
+
 use kimchi::{curve::KimchiCurve, proof::PointEvaluations};
+use kimchi_msm::{logup::LookupProof, LogupWitness, LookupTableID};
 use poly_commitment::{ipa::OpeningProof, PolyComm};
 
 use crate::interpreters::mips::{column::N_MIPS_SEL_COLS, witness::SCRATCH_SIZE};
 
-pub struct WitnessColumns<G, S> {
-    pub scratch: [G; SCRATCH_SIZE],
-    pub instruction_counter: G,
-    pub error: G,
+#[derive(Clone)]
+pub struct Lookup<F> {
+    pub m: Vec<F>,
+    pub f: Vec<F>,
+    pub t: F,
+}
+
+#[derive(Clone)]
+// TODO : Rename F and S
+pub struct WitnessColumns<F: Clone, S, ID: LookupTableID> {
+    pub scratch: [F; SCRATCH_SIZE],
+    pub instruction_counter: F,
+    pub error: F,
     pub selector: S,
+    pub lookup: BTreeMap<ID, Lookup<F>>,
+    pub lookup_agg: F,
 }
 
-pub struct ProofInputs<G: KimchiCurve> {
-    pub evaluations: WitnessColumns<Vec<G::ScalarField>, Vec<G::ScalarField>>,
+pub struct ProofInputs<G: KimchiCurve, ID: LookupTableID> {
+    pub evaluations: WitnessColumns<Vec<G::ScalarField>, Vec<G::ScalarField>, ID>,
+    pub logups: BTreeMap<ID, LogupWitness<G::ScalarField, ID>>,
 }
 
-impl<G: KimchiCurve> ProofInputs<G> {
+impl<G: KimchiCurve, ID: LookupTableID> ProofInputs<G, ID> {
     pub fn new(domain_size: usize) -> Self {
         ProofInputs {
             evaluations: WitnessColumns {
                 scratch: std::array::from_fn(|_| Vec::with_capacity(domain_size)),
                 instruction_counter: Vec::with_capacity(domain_size),
                 error: Vec::with_capacity(domain_size),
                 selector: Vec::with_capacity(domain_size),
+                lookup: BTreeMap::new(),
+                lookup_agg: Vec::with_capacity(domain_size),
             },
+            logups: BTreeMap::new(),
         }
     }
 }
 
 // FIXME: should we blind the commitment?
-pub struct Proof<G: KimchiCurve> {
-    pub commitments: WitnessColumns<PolyComm<G>, [PolyComm<G>; N_MIPS_SEL_COLS]>,
-    pub zeta_evaluations: WitnessColumns<G::ScalarField, [G::ScalarField; N_MIPS_SEL_COLS]>,
-    pub zeta_omega_evaluations: WitnessColumns<G::ScalarField, [G::ScalarField; N_MIPS_SEL_COLS]>,
+pub struct Proof<G: KimchiCurve, ID: LookupTableID> {
+    pub commitments: WitnessColumns<PolyComm<G>, [PolyComm<G>; N_MIPS_SEL_COLS], ID>,
+    pub zeta_evaluations: WitnessColumns<G::ScalarField, [G::ScalarField; N_MIPS_SEL_COLS], ID>,
+    pub zeta_omega_evaluations:
+        WitnessColumns<G::ScalarField, [G::ScalarField; N_MIPS_SEL_COLS], ID>,
```

Comment:
Should we reorganize the proof to have less fields, by putting all the commitments in one field (which can have more structure itself), and same for evals ?
This might also help in how we order polys in the opening proof
Just a suggestion, no strong opinion on my side